keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -159,69 +159,48 @@ take an aligned approach to adding apply to clients in a typesafe way.
+// associated with the actor or entity that is performing the apply and
+// is must be non-empty. force should be set to true if the apply should
+// re-qcquire conflicting fields owned by other fieldManagers.
+func(o ApplyOptions) NewApplyOptions(fieldManager string, force bool)
 ```
 
-<<[UNRESOLVED @jpbetz ]>>
+`ApplyOptions` is introduced primarily to allow us to include the
+`NewApplyOptions` function, and also to future proof the API.
 
-Should fieldManager be a parameter of Apply or should it be optional?
+The purpose of the `NewApplyOptions` function is to make it clear that
+fieldManager be a required parameter for Apply, and to encourage
+callers to decide if force should be set to true since, for
+controllers, it typically is.
 
-The main reason this KEP proposes fieldmanager be a required parameter
-for Apply is that some controllers have multiple code paths that
-update the same same object, but that update different field sets. If
-they used apply and the same field manager, fields could be
-accidentally removed or disowned.
+We do not provide a default for fieldManager since some controllers
+have multiple code paths that update the same same object, but that
+update different field sets. If they used apply and the same field
+manager, fields could be accidentally removed or disowned. We also
+make force a required field since it should typically be set to true
+for controllers but defaults to false.
 
 Create and Update methods default the fieldmanager to a user-agent
 based string. But unlike Apply, Updates succeed even if there are
 conflicts, so the fieldmanager name is almost entirely
 informational. For Apply, the fieldmanager name matters a lot more,
-and a user-agent string is almost certainly not a good default
-fieldmanager for Apply.
-
-If we require fieldManager, an alternative to making fieldManager a
-required parameter of `Apply` functions would be have a constructor
-for Apply options like `NewApplyOptions(fieldManager, force)` that
-makes the required field clear.
-
-<<[/UNRESOLVED]>>
-
-<<[UNRESOLVED @jpbetz ]>> 
-Should the Force field also be made a required field like fieldManager is?
-@lavalamp pointed out that user things typically want false but
-controller-things typically want true, so false isn't a great default.
-
-Also, can we structure this so that we don't have so many positional params?
-Moving two params from options to the top level positional params results in a
-longer function signature. Alternative proposed by @lavalamp was to do something
-more like this (ordering or parameters and function names TBD):
-
-```go
-client.Deployments().BeginApply("field-manager-name", options).Body(
-  appsv1apply.Deployment().
-    SetObjectMeta(&metav1apply.ObjectMeta()). ...
-).Apply(ctx)
-```
-
-<<[/UNRESOLVED]>>
-
-Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
-`PatchOptions`.
+and a user-agent string is not a good default for many use cases.
```

Comment:
Note that this is already the existing behavior from the apiserver's perspective. We don't default `FieldManager` for Apply requests (to the user-agent) and the field is required in t hat case.

Reply:
Added this to the KEP for clarity.